File: tests/cpp/test_image.cpp
#include <gtest/gtest.h>
#include "core/types/image.h"
#include "core/utils/image_io.h"
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <filesystem>

namespace fs = std::filesystem;

namespace hybrid_sfm {

class ImageTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Create test images
        test_image_path_ = "test_image.jpg";
        test_image_dir_ = "test_images";
        
        // Create a test image
        cv::Mat test_img(480, 640, CV_8UC3);
        cv::randu(test_img, cv::Scalar(0, 0, 0), cv::Scalar(255, 255, 255));
        cv::imwrite(test_image_path_, test_img);
        
        // Create test directory with multiple images
        fs::create_directory(test_image_dir_);
        for (int i = 0; i < 5; ++i) {
            cv::Mat img(240, 320, CV_8UC3);
            cv::randu(img, cv::Scalar(0, 0, 0), cv::Scalar(255, 255, 255));
            cv::imwrite(test_image_dir_ + "/image_" + std::to_string(i) + ".jpg", img);
        }
    }
    
    void TearDown() override {
        // Clean up test files
        fs::remove(test_image_path_);
        fs::remove_all(test_image_dir_);
    }
    
    std::string test_image_path_;
    std::string test_image_dir_;
};

// Test basic image loading
TEST_F(ImageTest, BasicLoading) {
    Image img(test_image_path_);
    
    EXPECT_EQ(img.getWidth(), 640);
    EXPECT_EQ(img.getHeight(), 480);
    EXPECT_EQ(img.getChannels(), 3);
    EXPECT_FALSE(img.getName().empty());
    EXPECT_EQ(img.getPath(), test_image_path_);
}

// Test lazy loading
TEST_F(ImageTest, LazyLoading) {
    Image img;
    bool loaded = img.load(test_image_path_, 
                          static_cast<int>(Image::LoadFlags::LAZY));
    
    EXPECT_TRUE(loaded);
    EXPECT_EQ(img.getPath(), test_image_path_);
    
    // Data should be loaded on demand
    EXPECT_EQ(img.getWidth(), 640);
    EXPECT_EQ(img.getHeight(), 480);
}

// Test pyramid generation
TEST_F(ImageTest, PyramidGeneration) {
    Image img(test_image_path_);
    img.buildPyramid(4, false);  // CPU pyramid
    
    const auto& pyramid = img.getPyramid();
    EXPECT_EQ(pyramid.getNumLevels(), 4);
    
    // Check pyramid dimensions
    cv::Mat level0 = pyramid.getLevel(0);
    cv::Mat level1 = pyramid.getLevel(1);
    
    EXPECT_EQ(level0.cols, 640);
    EXPECT_EQ(level0.rows, 480);
    EXPECT_EQ(level1.cols, 320);
    EXPECT_EQ(level1.rows, 240);
}

// Test GPU operations
TEST_F(ImageTest, GPUOperations) {
    if (cv::cuda::getCudaEnabledDeviceCount() == 0) {
        GTEST_SKIP() << "No CUDA devices available";
    }
    
    Image img(test_image_path_);
    
    // Test GPU upload
    img.uploadToGPU();
    EXPECT_TRUE(img.isOnGPU());
    
    // Test GPU pyramid
    img.buildPyramid(3, true);  // GPU pyramid
    
    // Test GPU grayscale conversion
    cv::Mat gray = img.getGray(true);
    EXPECT_EQ(gray.channels(), 1);
}

// Test memory management
TEST_F(ImageTest, MemoryManagement) {
    Image img(test_image_path_);
    
    size_t initial_memory = img.getMemoryUsage();
    EXPECT_GT(initial_memory, 0);
    
    // Upload to GPU should increase memory usage
    if (cv::cuda::getCudaEnabledDeviceCount() > 0) {
        img.uploadToGPU();
        size_t gpu_memory = img.getMemoryUsage();
        EXPECT_GT(gpu_memory, initial_memory);
        
        // Release GPU memory
        img.releaseGPUMemory();
        size_t after_release = img.getMemoryUsage();
        EXPECT_LT(after_release, gpu_memory);
    }
}

// Test image undistortion
TEST_F(ImageTest, Undistortion) {
    // Create a camera with some distortion
    std::vector<double> params = {500.0, 500.0, 320.0, 240.0, 0.1, -0.05, 0.01, 0.01};
    Camera camera(Camera::Model::OPENCV, 640, 480, params);
    
    Image img(test_image_path_);
    cv::Mat undistorted = img.getUndistorted(camera, false);
    
    EXPECT_EQ(undistorted.size(), img.getData().size());
    EXPECT_EQ(undistorted.type(), img.getData().type());
}

// Test metadata extraction
TEST_F(ImageTest, MetadataExtraction) {
    // This test would work with real images containing EXIF data
    Image img;
    img.load(test_image_path_, 
            static_cast<int>(Image::LoadFlags::EXTRACT_METADATA));
    
    const auto& metadata = img.getMetadata();
    // For synthetic test images, metadata will be empty
    EXPECT_TRUE(metadata.camera_make.empty() || !metadata.camera_make.empty());
}

// Test batch loading
TEST_F(ImageTest, BatchLoading) {
    ImageLoader::LoadOptions options;
    options.num_threads = 2;
    
    auto images = ImageLoader::loadDirectory(test_image_dir_, options, {});

    EXPECT_EQ(images.size(), 5);
    
    for (const auto& img : images) {
        EXPECT_NE(img, nullptr);
        EXPECT_EQ(img->getWidth(), 320);
        EXPECT_EQ(img->getHeight(), 240);
    }
}

// Test format detection
TEST_F(ImageTest, FormatDetection) {
    EXPECT_EQ(ImageLoader::detectFormat("test.jpg"), ImageFormat::JPEG);
    EXPECT_EQ(ImageLoader::detectFormat("test.png"), ImageFormat::PNG);
    EXPECT_EQ(ImageLoader::detectFormat("test.tiff"), ImageFormat::TIFF);
    EXPECT_EQ(ImageLoader::detectFormat("test.xyz"), ImageFormat::UNKNOWN);
}

// Test image sequence
TEST_F(ImageTest, ImageSequence) {
    ImageSequence seq(test_image_dir_);
    
    EXPECT_EQ(seq.getFrameCount(), 5);
    EXPECT_TRUE(seq.hasNext());
    
    // Test frame iteration
    int count = 0;
    while (seq.hasNext()) {
        auto frame = seq.getNextFrame();
        EXPECT_NE(frame, nullptr);
        count++;
    }
    EXPECT_EQ(count, 5);
    
    // Test reset
    seq.reset();
    EXPECT_EQ(seq.getCurrentIndex(), 0);
}

// Test image statistics
TEST_F(ImageTest, ImageStatistics) {
    Image img(test_image_path_);
    
    auto stats = image_utils::computeStats(img, false);
    
    // Check that statistics are computed
    EXPECT_GE(stats.mean[0], 0);
    EXPECT_LE(stats.mean[0], 255);
    EXPECT_GE(stats.min, 0);
    EXPECT_LE(stats.max, 255);
    EXPECT_FALSE(stats.histogram.empty());
}

// Test batch operations
TEST_F(ImageTest, BatchOperations) {
    auto images = ImageLoader::loadDirectory(test_image_dir_);
    
    // Test batch resize
    image_utils::batchResize(images, 160, false);
    
    for (const auto& img : images) {
        EXPECT_LE(img->getWidth(), 160);
        EXPECT_LE(img->getHeight(), 160);
    }
}

// Test PFM format
TEST_F(ImageTest, PFMFormat) {
    // Create a float image
    cv::Mat float_img(100, 100, CV_32FC1);
    cv::randu(float_img, 0.0f, 1.0f);
    
    // Save as PFM
    std::string pfm_path = "test.pfm";
    EXPECT_TRUE(ImageLoader::savePFM(pfm_path, float_img));
    
    // Load PFM
    cv::Mat loaded = ImageLoader::loadPFM(pfm_path);
    EXPECT_FALSE(loaded.empty());
    EXPECT_EQ(loaded.type(), CV_32FC1);
    EXPECT_EQ(loaded.size(), float_img.size());
    
    // Clean up
    fs::remove(pfm_path);
}

// Test memory limits
TEST_F(ImageTest, MemoryLimits) {
    // Set a low memory limit
    ImageLoader::setMaxMemoryUsage(1024 * 1024);  // 1MB
    
    // Try to load multiple large images
    std::vector<std::string> paths;
    for (int i = 0; i < 10; ++i) {
        paths.push_back(test_image_path_);
    }
    
    ImageLoader::LoadOptions options;
    options.skip_corrupted = true;
    
    auto images = ImageLoader::loadBatch(paths, options);
    
    // Some images might not load due to memory limit
    EXPECT_LE(images.size(), paths.size());
    
    // Reset memory limit
    ImageLoader::setMaxMemoryUsage(4ULL * 1024 * 1024 * 1024);
}

// Test thread safety
TEST_F(ImageTest, ThreadSafety) {
    Image img(test_image_path_);
    
    // Concurrent operations
    std::vector<std::thread> threads;
    
    for (int i = 0; i < 4; ++i) {
        threads.emplace_back([&img, i]() {
            if (i % 2 == 0) {
                cv::Mat gray = img.getGray(false);
            } else {
                cv::Mat resized = img.getResized(160, 120, false);
            }
        });
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    // Should complete without issues
    EXPECT_TRUE(img.isThreadSafe());
}

} // namespace hybrid_sfm

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: tests/python/test_image.py
#!/usr/bin/env python3
"""
Python unit tests for the enhanced Image class
"""

import unittest
import numpy as np
import cv2
import tempfile
import shutil
import os
from pathlib import Path

# Import the hybrid_sfm module (assuming it's properly built and installed)
import hybrid_sfm


class TestImage(unittest.TestCase):
    """Test cases for Image class functionality"""
    
    def setUp(self):
        """Create test images and directories"""
        self.test_dir = tempfile.mkdtemp()
        
        # Create test image
        self.test_image = np.random.randint(0, 255, (480, 640, 3), dtype=np.uint8)
        self.test_image_path = os.path.join(self.test_dir, "test.jpg")
        cv2.imwrite(self.test_image_path, self.test_image)
        
        # Create multiple test images for batch operations
        self.num_test_images = 5
        for i in range(self.num_test_images):
            img = np.random.randint(0, 255, (240, 320, 3), dtype=np.uint8)
            path = os.path.join(self.test_dir, f"image_{i:03d}.jpg")
            cv2.imwrite(path, img)
    
    def tearDown(self):
        """Clean up test files"""
        shutil.rmtree(self.test_dir)
    
    def test_basic_loading(self):
        """Test basic image loading"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        self.assertEqual(img.width, 640)
        self.assertEqual(img.height, 480)
        self.assertEqual(img.channels, 3)
        self.assertIsNotNone(img.name)
        self.assertEqual(img.path, self.test_image_path)
    
    def test_lazy_loading(self):
        """Test lazy loading functionality"""
        img = hybrid_sfm.Image()
        # Note: The C++ enum is accessed via the class
        flags = hybrid_sfm.Image.LoadFlags.LAZY
        
        success = img.load(self.test_image_path, flags)
        self.assertTrue(success)
        
        # Data should be loaded on first access
        self.assertEqual(img.width, 640)
        self.assertEqual(img.height, 480)
    
    def test_numpy_interface(self):
        """Test numpy array conversion"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        # Get as numpy array
        arr = img.to_numpy()
        self.assertIsInstance(arr, np.ndarray)
        self.assertEqual(arr.shape, (480, 640, 3))
        
        # Create from numpy
        new_img = hybrid_sfm.Image.from_numpy(arr, "from_numpy")
        self.assertEqual(new_img.width, 640)
        self.assertEqual(new_img.height, 480)
        self.assertEqual(new_img.name, "from_numpy")
    
    def test_pyramid_generation(self):
        """Test image pyramid generation"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        # Build pyramid
        num_levels = 4
        img.build_pyramid(num_levels, use_gpu=False)
        
        pyramid = img.pyramid
        self.assertEqual(pyramid.num_levels, num_levels)
        
        # Check level dimensions
        for i in range(num_levels):
            level = pyramid.get_level(i)
            expected_width = 640 // (2**i)
            expected_height = 480 // (2**i)
            self.assertEqual(level.shape[1], expected_width)
            self.assertEqual(level.shape[0], expected_height)
    
    @unittest.skipIf(not hybrid_sfm.has_cuda(), "CUDA not available")
    def test_gpu_operations(self):
        """Test GPU operations if CUDA is available"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        # Upload to GPU
        img.upload_to_gpu()
        self.assertTrue(img.is_on_gpu())
        
        # GPU grayscale conversion
        gray = img.get_gray(use_gpu=True)
        self.assertEqual(gray.ndim, 2)
        
        # GPU resize
        resized = img.get_resized(320, 240, use_gpu=True)
        self.assertEqual(resized.shape[:2], (240, 320))
    
    def test_memory_management(self):
        """Test memory usage tracking"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        initial_memory = img.memory_usage
        self.assertGreater(initial_memory, 0)
        
        if hybrid_sfm.has_cuda():
            # GPU upload should increase memory
            img.upload_to_gpu()
            gpu_memory = img.memory_usage
            self.assertGreater(gpu_memory, initial_memory)
            
            # Release GPU memory
            img.release_gpu_memory()
            after_release = img.memory_usage
            self.assertLess(after_release, gpu_memory)
    
    def test_undistortion(self):
        """Test image undistortion"""
        camera = hybrid_sfm.Camera(
            hybrid_sfm.Camera.Model.OPENCV,
            640, 480,
            [500.0, 500.0, 320.0, 240.0, 0.1, -0.05, 0.01, 0.01]
        )
        
        img = hybrid_sfm.Image(self.test_image_path)
        undistorted = img.get_undistorted(camera, use_gpu=False)
        
        self.assertEqual(undistorted.shape, (480, 640, 3))
    
    def test_metadata(self):
        """Test metadata extraction"""
        img = hybrid_sfm.Image()
        flags = hybrid_sfm.Image.LoadFlags.EXTRACT_METADATA
        img.load(self.test_image_path, flags)
        
        metadata = img.metadata
        self.assertIsInstance(metadata.camera_make, str)
    
    def test_image_stats(self):
        """Test image statistics computation"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        stats = hybrid_sfm.compute_image_stats(img, use_gpu=False)
        
        self.assertIn('mean', stats)
        self.assertIn('stddev', stats)
        self.assertIn('min', stats)
        self.assertIn('max', stats)
        self.assertIn('histogram', stats)
        
        self.assertGreaterEqual(stats['min'], 0)
        self.assertLessEqual(stats['max'], 255)
    
    def test_batch_loading(self):
        """Test batch image loading"""
        options = hybrid_sfm.LoadOptions()
        options.num_threads = 2
        
        images = hybrid_sfm.load_images_from_directory(
            self.test_dir, options
        )
        
        self.assertEqual(len(images), self.num_test_images + 1)
        
        for img in images:
            self.assertIsNotNone(img)
            self.assertGreater(img.width, 0)
            self.assertGreater(img.height, 0)
    
    def test_image_sequence(self):
        """Test image sequence handling"""
        seq = hybrid_sfm.ImageSequence(self.test_dir)
        
        self.assertEqual(seq.frame_count, self.num_test_images + 1)
        self.assertTrue(seq.has_next())
        
        count = 0
        while seq.has_next():
            frame = seq.get_next_frame()
            self.assertIsNotNone(frame)
            count += 1
        
        self.assertEqual(count, self.num_test_images + 1)
        seq.reset()
        self.assertEqual(seq.current_index, 0)
    
    def test_format_detection(self):
        """Test image format detection"""
        self.assertEqual(hybrid_sfm.detect_image_format("test.jpg"), hybrid_sfm.ImageFormat.JPEG)
        self.assertEqual(hybrid_sfm.detect_image_format("test.png"), hybrid_sfm.ImageFormat.PNG)
        self.assertEqual(hybrid_sfm.detect_image_format("test.xyz"), hybrid_sfm.ImageFormat.UNKNOWN)
    
    def test_pfm_format(self):
        """Test PFM (Portable Float Map) format"""
        float_img = np.random.rand(100, 100).astype(np.float32)
        pfm_path = os.path.join(self.test_dir, "test.pfm")
        
        success = hybrid_sfm.save_pfm(pfm_path, float_img)
        self.assertTrue(success)
        
        loaded = hybrid_sfm.load_pfm(pfm_path)
        self.assertIsNotNone(loaded)
        self.assertEqual(loaded.shape, float_img.shape)
        self.assertEqual(loaded.dtype, np.float32)
    
    def test_batch_operations(self):
        """Test batch image operations"""
        images = hybrid_sfm.load_images_from_directory(self.test_dir)
        
        # Batch resize
        hybrid_sfm.batch_resize(images, max_dimension=160, use_gpu=False)
        
        for img in images:
            self.assertLessEqual(img.width, 160)
            self.assertLessEqual(img.height, 120) # 320x240 -> 160x120 or 640x480 -> 160x120
        
        # Batch normalization (check that it runs without error)
        try:
            hybrid_sfm.batch_normalize(images, mean=128.0, std=64.0, use_gpu=False)
        except Exception as e:
            self.fail(f"batch_normalize raised an exception: {e}")

    def test_batch_undistort(self):
        """Test batch image undistortion"""
        images = hybrid_sfm.load_images_from_directory(self.test_dir)
        camera = hybrid_sfm.Camera(
            hybrid_sfm.Camera.Model.OPENCV,
            320, 240, # For the smaller images in the batch
            [200.0, 200.0, 160.0, 120.0, 0.1, -0.05, 0.01, 0.01]
        )
        
        # Find a small image to test against
        small_image = next((img for img in images if img.width == 320), None)
        self.assertIsNotNone(small_image)
        
        original_data = small_image.to_numpy().copy()

        # Apply batch undistort
        hybrid_sfm.batch_undistort(images, camera, use_gpu=False)

        # Check that the data has changed
        undistorted_data = small_image.to_numpy()
        self.assertFalse(np.array_equal(original_data, undistorted_data))
        self.assertEqual(original_data.shape, undistorted_data.shape)

    def test_color_conversions(self):
        """Test color space conversions"""
        img = hybrid_sfm.Image(self.test_image_path)
        
        gray = img.get_gray(use_gpu=False)
        self.assertEqual(len(gray.shape), 2)
        
        bgr = img.to_numpy()
        rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)
        
        img_rgb = hybrid_sfm.Image.from_numpy(rgb, "rgb_image")
        self.assertEqual(img_rgb.width, img.width)
    
    def test_memory_limits(self):
        """Test memory limit enforcement"""
        hybrid_sfm.set_max_memory_usage(1024 * 1024)  # 1MB
        
        paths = [self.test_image_path] * 10
        options = hybrid_sfm.LoadOptions()
        options.skip_corrupted = True
        
        images = hybrid_sfm.load_images_batch(paths, options)
        
        self.assertLessEqual(len(images), len(paths))
        hybrid_sfm.set_max_memory_usage(4 * 1024 * 1024 * 1024)  # 4GB
    
    def test_thread_safety(self):
        """Test thread-safe operations"""
        import threading
        
        img = hybrid_sfm.Image(self.test_image_path)
        errors = []
        
        def worker(img, errors):
            try:
                _ = img.get_gray(use_gpu=False)
                _ = img.get_resized(160, 120, use_gpu=False)
                _ = hybrid_sfm.compute_image_stats(img)
            except Exception as e:
                errors.append(e)
        
        threads = []
        for _ in range(4):
            t = threading.Thread(target=worker, args=(img, errors))
            t.start()
            threads.append(t)
        
        for t in threads:
            t.join()
        
        self.assertEqual(len(errors), 0)


class TestImagePerformance(unittest.TestCase):
    """Performance tests for image operations"""
    
    def setUp(self):
        """Create larger test images for performance testing"""
        self.test_dir = tempfile.mkdtemp()
        
        self.large_image = np.random.randint(0, 255, (2048, 2048, 3), dtype=np.uint8)
        self.large_image_path = os.path.join(self.test_dir, "large.jpg")
        cv2.imwrite(self.large_image_path, self.large_image)
    
    def tearDown(self):
        """Clean up test files"""
        shutil.rmtree(self.test_dir)
    
    @unittest.skipIf(not hybrid_sfm.has_cuda(), "CUDA not available")
    def test_gpu_speedup(self):
        """Test GPU acceleration speedup"""
        img = hybrid_sfm.Image(self.large_image_path)
        
        import time
        
        start = time.time()
        _ = img.get_gray(use_gpu=False)
        cpu_time = time.time() - start
        
        start = time.time()
        img.upload_to_gpu()
        _ = img.get_gray(use_gpu=True)
        gpu_time = time.time() - start
        
        print(f"CPU time: {cpu_time:.4f}s, GPU time: {gpu_time:.4f}s")
    
    def test_pyramid_performance(self):
        """Test pyramid generation performance"""
        img = hybrid_sfm.Image(self.large_image_path)
        
        import time
        
        start = time.time()
        img.build_pyramid(5, use_gpu=hybrid_sfm.has_cuda())
        pyramid_time = time.time() - start
        
        print(f"Pyramid generation time: {pyramid_time:.4f}s")
        self.assertLess(pyramid_time, 1.0)


if __name__ == '__main__':
    unittest.main()
